pkg/hydrate/controller.go

```diff
@@ -327,60 +337,82 @@ func deleteErrorFile(file string) error {
+			// ready, so return the error directly without retry.
+			return fmt.Errorf("error in the %s container: %s", containerName, string(content))
+		default:
+			// The source error file doesn't exist => *-sync is ready without any errors.
+		}
 
-	gitDir, err := sourceRevDir.EvalSymlinks()
-	if err != nil {
-		return "", "", status.SourceError.Sprintf("unable to evaluate the source link %s", sourceRevDir).Wrap(err).Build()
-	}
+		gitDir, err := sourceRevDir.EvalSymlinks()
+		if err != nil {
+			// `sourceRevDir` points to the directory with commit SHA that holds the
+			// checked-out files. It can't be evaluated probably because *-sync
+			// container is ready, but hasn't finished creating the symlink yet, so
+			// retry until the symlink is created.
+			return util.NewRetriableError(fmt.Errorf("failed to evaluate the source rev directory %q: %v", sourceRevDir, err))
+		}
 
-	commit := filepath.Base(gitDir.OSPath())
-
-	// The hydration controller might pull remote Helm charts locally, which makes the source directory dirty.
-	// Hence, we don't check if the source directory is clean before the hydration.
-	// The assumption is that customers should have limited access to manually modify the source configs.
-	// For the local Helm charts pulled by the Helm inflator, the entire hydrated directory
-	// will blow away when new commits come in.
-	// If the commit hash is not changed, the hydration will be skipped.
-	// Therefore, it is relatively safe to keep the Helm charts local in the source directory.
-	relSyncDir := gitDir.Join(syncDir)
-	sourceDir, err := relSyncDir.EvalSymlinks()
-	if err != nil {
-		return commit, "", status.PathWrapError(
-			errors.Wrap(err, "evaluating symbolic link to policy sourceRoot"), relSyncDir.OSPath())
-	}
-	return commit, sourceDir, nil
+		commit = filepath.Base(gitDir.OSPath())
+
+		// The hydration controller might pull remote Helm charts locally, which makes the source directory dirty.
+		// Hence, we don't check if the source directory is clean before the hydration.
+		// The assumption is that customers should have limited access to manually modify the source configs.
+		// For the local Helm charts pulled by the Helm inflator, the entire hydrated directory
+		// will blow away when new commits come in.
+		// If the commit hash is not changed, the hydration will be skipped.
+		// Therefore, it is relatively safe to keep the Helm charts local in the source directory.
+		relSyncDir := gitDir.Join(syncDir)
+		sourceDir, err = relSyncDir.EvalSymlinks()
+		if err != nil {
+			// `relSyncDir` points to the source config directory. Now the commit SHA
+			// has been checked out, but it fails to evaluate the symlink with the
+			// config directory, which indicates a misconfiguration of `spec.git.dir`,
+			// `spec.oci.dir`, or `spec.helm.dir`, so return the error without retry.
+			return fmt.Errorf("failed to evaluate the config directory %q: %w", relSyncDir.OSPath(), err)
```

Comment:
It looks like this error is changed from a `PathError` to a `SourceError`. Is that intentional?

Reply:
Yes. That's intentional. I think it was wrong initially. `PathError` is used for invalid YAML configuration and special characters in the path: https://cloud.google.com/anthos-config-management/docs/reference/errors#knv2001.

It makes more sense to classify the misconfiguration of `spec.git.dir` as a source error.